Create migration guide from AI SDK to TanStack

tannerlinsley opened the pull request:
Add comprehensive migration guide covering:
- Package installation differences
- Server-side API migration (streamText -> chat)
- Client-side useChat hook differences
- Isomorphic tool system migration
- Provider adapter changes (OpenAI, Anthropic, Gemini)
- Streaming response formats
- Multimodal content handling
- Type safety enhancements
- Complete before/after code examples

## 🎯 Changes

<!-- What changes are made in this PR? Describe the change and its motivation. -->

## ✅ Checklist

- [ ] I have followed the steps in the [Contributing guide](https://github.com/TanStack/ai/blob/main/CONTRIBUTING.md).
- [ ] I have tested this code locally with `pnpm run test:pr`.

## 🚀 Release Impact

- [ ] This change affects published code, and I have generated a [changeset](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md).
- [ ] This change is docs/CI/dev-only (no release).


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Documentation**
  * Added a comprehensive migration guide for moving from Vercel AI SDK to TanStack AI, covering installation, API/config mappings, server & client migration patterns, streaming/structured outputs, tool/function-calling, middleware, observability, typed usage, and examples.
  * Guide is now listed under the Migration section in the docs navigation.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

claude committed:
docs: add migration guide from Vercel AI SDK to TanStack AI

Add comprehensive migration guide covering:
- Package installation differences
- Server-side API migration (streamText -> chat)
- Client-side useChat hook differences
- Isomorphic tool system migration
- Provider adapter changes (OpenAI, Anthropic, Gemini)
- Streaming response formats
- Multimodal content handling
- Type safety enhancements
- Complete before/after code examples

tannerlinsley commented on docs/guides/migration-from-vercel-ai.md:
````diff
@@ -0,0 +1,937 @@
+### Before (Vercel AI SDK)
+
+```bash
+npm install ai @ai-sdk/openai @ai-sdk/anthropic
````
@claude fix this

AlemTuzlak replied:
Fixed in ebd689c1. Added @ai-sdk/react to the v5+ install command and updated the Quick Reference table so every framework row points at its current @ai-sdk/* package (@ai-sdk/react, @ai-sdk/vue, @ai-sdk/solid, @ai-sdk/svelte). Added a short note that v4's ai/react subpath was replaced by the dedicated package in v5.

AlemTuzlak committed:
Merge branch 'main' into claude/migration-guide-versace-tanstack-NqpkX

AlemTuzlak committed:
Merge branch 'main' into claude/migration-guide-versace-tanstack-NqpkX

jherr committed:
Merge branch 'main' into claude/migration-guide-versace-tanstack-NqpkX

AlemTuzlak committed:
docs: address review feedback on Vercel AI migration guide

- Installation: add @ai-sdk/react to v5+ deps and update quick-reference
  table to show the v5 framework package names
- System prompts: use root-level systemPrompts: [...] instead of
  prepending a system message to the messages array (verified against
  packages/typescript/ai/src/types.ts)
- useChat API table: rewrite against current Vercel AI SDK v5+ API
  (sendMessage, status, regenerate, DefaultChatTransport) so the
  comparison is accurate rather than mixing v4/v5
- MessagePart: expand to full discriminated union with real field names
  (arguments/input/approval on tool-call, content on tool-result) and
  real ToolCallState values
- Fix nonexistent toStreamResponse references -> toServerSentEventsResponse
  (and add toHttpResponse where appropriate)
- Fix AbortController section heading (h4 -> h3, resolves MD001)
- Update tool schema section to note parameters -> inputSchema rename
  in AI SDK v5
- Tighten tool approval example with optional chaining and a note on
  arguments vs parsed input

AlemTuzlak commented on docs/guides/migration-from-vercel-ai.md:
````diff
@@ -0,0 +1,937 @@
+```typescript
+const stream = chat({
+  adapter: openaiText('gpt-4o'),
+  messages: [
````
Fixed in ebd689c1. You were right — chat() takes `systemPrompts: string[]` at the root. Verified against packages/typescript/ai/src/types.ts:572 and the chat options at packages/typescript/ai/src/activities/chat/index.ts. Also updated the final complete example (CodeRabbit flagged the same issue at line 928). The System Messages section now shows the multi-prompt form too, since systemPrompts is an array.

AlemTuzlak commented on docs/guides/migration-from-vercel-ai.md:
```diff
@@ -0,0 +1,937 @@
+| Vercel AI SDK | TanStack AI | Notes |
+|--------------|-------------|-------|
+| `api: '/api/chat'` | `connection: fetchServerSentEvents('/api/chat')` | Explicit connection adapter |
+| `input`, `handleInputChange` | Manage state yourself | More control, less magic |
```
Fixed in ebd689c1. You were right to flag this — the useChat API comparison was mixing Vercel v4 (`input`/`handleInputChange`/`handleSubmit`) with the TanStack headless-state philosophy, which made the claim "Same pattern" for `append/reload/stop/setMessages` misleading. Vercel AI SDK v5+ already moved to headless state (sendMessage, status, regenerate, DefaultChatTransport), so I rewrote the whole table against the current v5+ surface and the actually-exported useChat return from packages/typescript/ai-react/src/use-chat.ts (sendMessage, append, reload, stop, isLoading, setMessages, addToolResult, addToolApprovalResponse). The basic useChat example above the table was updated to match.

AlemTuzlak commented:
Addressed the review feedback in ebd689c1. Summary of what changed and why, after verifying every API claim against the current TanStack AI source and the Vercel AI SDK v5+ docs:

**Human feedback (Alem + Tanner)**
- **Line 39 — `ai/react` missing.** Added `@ai-sdk/react` to the v5+ install command. The Quick Reference table rows for framework hooks now point at the actual v5+ packages (`@ai-sdk/react`, `@ai-sdk/vue`, `@ai-sdk/solid`, `@ai-sdk/svelte`) with a note about the v4 → v5 subpath move.
- **Line 150 — System prompt wrong.** `chat()` takes `systemPrompts: string[]` at the root (verified in `packages/typescript/ai/src/types.ts:572`). Replaced the prepended-system-message pattern in both the System Messages section and the Complete Example. Added an example of multi-prompt composition since it's an array.
- **Line 231 — "This seems hallucinated?".** The useChat API table was comparing Vercel v4 magic (`input`/`handleInputChange`/`handleSubmit`) with TanStack's headless model, then claiming `append/reload/stop/setMessages` were "Same pattern". Vercel v5+ already went headless, so the table is rewritten against the v5+ surface and the actual TanStack `useChat` return from `packages/typescript/ai-react/src/use-chat.ts` (sendMessage, append, reload, stop, isLoading, setMessages, addToolResult, addToolApprovalResponse).

**CodeRabbit feedback**
- **MessagePart type incomplete (line 266).** Expanded to the real discriminated union from `packages/typescript/ai/src/types.ts:248-287` and `packages/typescript/ai-client/src/types.ts:32-132`: `text`, `tool-call` (with `id`, `name`, `arguments` JSON string, optional parsed `input`, `state`, `approval`, `output`), `tool-result` (with `toolCallId`, `content`, `state`, `error`), and `thinking`. Documented real `ToolCallState` values (`awaiting-input | input-streaming | input-complete | approval-requested | approval-responded`) and `ToolResultState` values. Added a note clarifying TanStack does not have separate `reasoning`/`source-url`/`source-document`/`file` part types.
- **Heading level jump (line 614, MD001).** AbortController subsections now use `###` like every other section.
- **systemPrompts in final example (line 928).** Fixed in the same pass as the System Messages section.

**Other corrections found while verifying**
- `toStreamResponse` does not exist in `@tanstack/ai`. The real exports are `streamToText`, `toServerSentEventsStream`/`toServerSentEventsResponse`, `toHttpStream`/`toHttpResponse` (see `packages/typescript/ai/src/stream-to-response.ts` and `packages/typescript/ai/src/index.ts:56-60`). Replaced every `toStreamResponse` reference, including in the Streaming Responses section, the Complete Example, and the Basic Text Generation example. Also added `toHttpResponse` as a documented option.
- Before-examples for Vercel now reflect v5+ (`@ai-sdk/react`, `DefaultChatTransport`, `sendMessage`, `status`, `maxOutputTokens`, `providerOptions`, `inputSchema` on `tool()`) with a v4 note where naming changed.
- Tool approval example tightened: uses optional chaining, narrows with `part.approval`, and notes the difference between raw `arguments` (streaming JSON string) and parsed `input`.

I already pushed the commit; the PR is green for re-review.

AlemTuzlak committed:
Merge remote-tracking branch 'origin/main' into pr-179

# Conflicts:
#	docs/tools/migration-from-vercel-ai.md

AlemTuzlak committed:
docs: expand Vercel AI SDK migration guide with full option + v6 coverage

- Add exhaustive streamText -> chat() option mapping table covering
  every AI SDK v6 parameter (tools, toolChoice, activeTools, stopWhen,
  prepareStep, experimental_transform/context/telemetry/repairToolCall,
  all sampling controls, abort, headers, providerOptions -> modelOptions)
- Add streamText result -> TanStack equivalent table (textStream,
  fullStream, text, usage, finishReason, steps, toUIMessageStreamResponse,
  pipeTextStreamToResponse, consumeStream, etc.)
- Expand Generation Options with topK/presence/frequency/seed/stop under
  modelOptions, clarify flat typed modelOptions vs provider-keyed
  providerOptions
- New section: Structured Output (generateObject / streamObject / v6
  Output.object) -> outputSchema on chat(); notes on Standard Schema
  libraries, provider strategies, and the current gap for partial
  object streaming
- New section: Agent Loop Control — stopWhen / hasToolCall / stepCountIs
  mapped to maxIterations / untilFinishReason / combineStrategies, and
  prepareStep mapped to middleware onConfig/onIteration
- New section: Middleware — wrapLanguageModel + experimental_transform
  mapped to a single ChatMiddleware array; full hook inventory;
  toolCacheMiddleware usage; common-pattern mapping table
- New section: Observability — where to plug logging/metrics/tracing
- Update generateText coverage to chat({ stream: false }) returning a
  real Promise<string> (not just streamToText)
- Update Tool Approval "Before" to show AI SDK v6's native needsApproval
  + sendAutomaticallyWhen flow; the two APIs are now symmetric
- Reframe "Removed Features" -> "Features Not Yet Covered" and scope
  it to embeddings, partial-object streaming, built-in retries/timeouts
- Update frontmatter for the docs/migration/ location (order, description,
  keywords); fix cross-links to the new directory layout
  (../advanced/middleware, ../chat/structured-outputs, etc.)

AlemTuzlak committed:
docs: round-1 CR fixes on Vercel AI migration guide

Factual corrections verified against source:
- Multimodal image source shape uses { type: 'url'|'data', value, mimeType }
  not { url, base64, mediaType } (types.ts:142-183)
- toolCacheMiddleware is exported from @tanstack/ai/middlewares, not the
  root (packages/typescript/ai/src/middlewares/index.ts)
- toolDefinition({ description }) is required; add it to the two doc
  examples that were missing it (tool-definition.ts:31)
- stream() connection adapter factory is (messages, data?) with no
  signal arg; rewrite custom-adapter example (connection-adapters.ts:441)

AI SDK v6 accuracy:
- addToolResult -> addToolOutput (v6 rename)
- experimental_output -> output (de-experimentalized)
- Soften "replaced" claim about generateObject/streamObject — they are
  deprecated, not removed
- Vercel addToolApprovalResponse row: v6 has this; replace "N/A"
- First Basic Text Generation Before example now uses v5+ API
  (convertToModelMessages + toUIMessageStreamResponse) with a v4
  toDataStreamResponse callout

Consistency:
- Agent-loop tables reconciled: only one truly built-in strategy
  (maxIterations / untilFinishReason / combineStrategies); hasToolCall
  requires a custom AgentLoopStrategy. Both tables now agree.
- prepareStep Before/After actually demonstrates equivalent behavior:
  Before shows step-level config tweak, After uses onConfig;
  mid-loop model switching split into its own subsection with the
  two-chat pattern the prose describes
- Message Structure section qualifies that ToolCallPart.input is the
  ai-client projection (server-side reads arguments directly)
- toHttpStream/Response comment in client connection example clarified
- Complete Example clarifies why convertToModelMessages disappears in
  the After (chat() accepts UI messages directly)
- clientTools() auto-execution comment expanded to state that no
  onToolCall/addToolOutput call is needed
- Anchor slug for Structured Output simplified to #structured-output

Rot hygiene:
- "current releases" removed from v5/v6 note
- "Every option" softened to "Options accepted ... as of AI SDK v6"
- "now expose" / "AI SDK v6 offers" / "v6 consolidated" reworded to
  avoid tense decay across future releases

AlemTuzlak commented:
Ran the CR loop on the expanded guide (3 specialized review agents, 2 rounds).

**Round 1 → 15 actionable findings, all fixed in 711dc15b**

_Factual corrections against the TanStack AI source:_
- Multimodal image source shape is `{ type: 'url'|'data', value, mimeType }`, not `{ url/base64/mediaType }` (verified at `packages/typescript/ai/src/types.ts:142-183`).
- `toolCacheMiddleware` is exported from `@tanstack/ai/middlewares`, not `@tanstack/ai` (subpath export in `packages/typescript/ai/package.json`).
- `toolDefinition({ description })` is required, not optional (`tool-definition.ts:31`) — two examples were missing it.
- `stream()` connection-adapter factory takes `(messages, data?)`, no `signal` (`connection-adapters.ts:441`).
- `ToolCallPart.input` (parsed) only exists on the ai-client projection; core reads `arguments` (raw JSON). Message Structure section now scoped to ai-client.

_AI SDK v6 accuracy:_
- `addToolResult` → `addToolOutput` (v6 rename, verified at [chatbot-tool-usage](https://ai-sdk.dev/docs/ai-sdk-ui/chatbot-tool-usage)).
- `experimental_output` → `output` (de-experimentalized; verified at [generate-text reference](https://ai-sdk.dev/docs/reference/ai-sdk-core/generate-text)).
- Softened "replaced" claim about `generateObject`/`streamObject` — they're deprecated but still present in v6.
- Vercel `addToolApprovalResponse` row was marked `N/A` in the useChat table but used in the Tool Approval Before example; table fixed.
- First Basic Text Generation Before now uses v5+ API (`convertToModelMessages` + `toUIMessageStreamResponse()`) with a v4 note.

_Internal consistency:_
- Agent-loop tables reconciled: `maxIterations` / `untilFinishReason` / `combineStrategies` are the built-in strategies; `hasToolCall` needs a one-line custom strategy. Both tables now say the same thing.
- `prepareStep` Before/After rewritten so the two snippets are actually equivalent. Mid-loop model switching split into a subsection with the two-chat pattern the prose describes.
- Complete Example explains why `convertToModelMessages` disappears in the After.
- `clientTools()` \"// Auto-executed\" comment expanded to make explicit that no `onToolCall`/`addToolOutput` handler is needed.
- `toHttpStream`/`toHttpResponse` pairing clarified on the client side.
- Anchor slug for Structured Output simplified to \`#structured-output\` (avoids slugifier ambiguity with punctuation-heavy heading).

_Rot hygiene:_
- \"current releases\" / \"now expose\" / \"AI SDK v6 offers\" / \"v6 consolidated\" reworded so the guide doesn't silently decay across future AI SDK releases.

**False positive from the review pass:** one agent flagged \"Embeddings: Not Yet Covered\" as factually wrong because CLAUDE.md mentioned `openaiEmbed` / `embedding()`. I verified nothing with those names exists in `packages/typescript/ai/src/` or `packages/typescript/ai-openai/src/` today. Left the doc claim as-is.

**Round 2 → 0 findings.** Each round-1 fix verified against the source; fresh scan found no regressions. Ready for re-review.

AlemTuzlak committed:
docs: use toServerSentEventsResponse/toHttpResponse init options

Both helpers accept ResponseInit & { abortController }, so custom headers,
status, and cancellation flow through the helpers directly. Drop the
hand-rolled `new Response(toServerSentEventsStream(...), { headers: {...} })`
example and keep the raw stream helpers only for the genuine "pipe elsewhere"
case.
